src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,131 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private int size = 0;
+    private int maxSize = 0;
+    private Object[] elementData;
+
     @Override
     public void add(T value) {
-
+        if (this.isEmpty()) {
+            elementData = new Object[DEFAULT_SIZE];
+            maxSize = DEFAULT_SIZE;
+        } else {
+            if (this.size == maxSize) {
+                maxSize = maxSize + maxSize / 2;
+                Object[] oldElementData = elementData;
+                elementData = new Object[maxSize];
+                System.arraycopy(oldElementData, 0, elementData, 0, size);
+            }
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if ((index < 0) || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Out of bound exception");
+        }
+        if (index <= size) {
+            if (size != 0) {
+                if (this.size == maxSize) {
+                    maxSize = maxSize + maxSize / 2;
+                }
+                Object[] oldElementsData = new Object[maxSize];
+                System.arraycopy(elementData, 0, oldElementsData, 0, size);
+                elementData = new Object[size + 1];
+                System.arraycopy(oldElementsData, 0, elementData, 0, index);
+                elementData[index] = value;
+                System.arraycopy(oldElementsData, index, elementData, index + 1, size - index);
+                size++;
+            } else {
+                this.add(value);
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            for (int i = 0; i < list.size(); i++) {
+                this.add(list.get(i));
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if ((index < 0) || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Out of bound exception");
+        }
+        Objects.checkIndex(index, size);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if ((index < 0) || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Out of bound exception");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if ((index < 0) || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Out of bound exception");
+        }
+        Objects.checkIndex(index, size);
+        final Object[] es = elementData;
+
+        T oldValue = (T) es[index];
+        fastRemove(es, index);
+
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
```

Comment:
just check each element in loop for equals and remove use your method remove(index)